Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm

```diff
@@ -5558,99 +5557,57 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
-        Ref view = scroller->checkedView()->frameView();
-        IntRect scrollerClipRectInContent;
-        std::optional<ScrollingNodeID> enclosingScrollingNodeID;
-        if (CheckedPtr renderView = dynamicDowncast<RenderView>(*scroller)) {
-            if (renderView->protectedDocument()->isTopDocument())
-                break;
+    if (!state.isContentEditable && selection.isCaret())
+        return;
 
-            scrollerClipRectInContent = view->visibleContentRect();
-            enclosingScrollingNodeID = view->scrollingNodeID();
-        } else if (CheckedPtr layer = scroller->layer()) {
-            CheckedPtr scrollableArea = layer->scrollableArea();
-            if (!scrollableArea)
-                continue;
+    RefPtr startContainer = selection.start().containerNode();
+    if (!startContainer)
+        return;
 
-            scrollerClipRectInContent = scroller->absoluteBoundingBoxRect();
-            enclosingScrollingNodeID = scrollableArea->scrollingNodeID();
-        }
+    RefPtr endContainer = selection.end().containerNode();
+    if (!endContainer)
+        return;
 
-        if (callback(WTFMove(enclosingScrollingNodeID), view->contentsToRootView(scrollerClipRectInContent)) == IterationStatus::Done)
-            break;
-    }
-}
+    CheckedPtr startRenderer = startContainer->renderer();
+    if (!startRenderer)
+        return;
 
-void WebPage::computeSelectionClipRectAndEnclosingScroller(EditorState& state, const VisibleSelection& selection, const IntRect& editableRootBounds) const
-{
-    if (selection.isNoneOrOrphaned())
+    CheckedPtr endRenderer = startContainer->renderer();
+    if (!endRenderer)
         return;
 
-    if (!state.isContentEditable && selection.isCaret())
+    CheckedPtr startEnclosingLayer = startRenderer->enclosingLayer();
+    if (!startEnclosingLayer)
         return;
 
-    if (!m_selectionHonorsOverflowScrolling) {
-        state.visualData->selectionClipRect = editableRootBounds;
+    CheckedPtr endEnclosingLayer = endRenderer->enclosingLayer();
+    if (!endEnclosingLayer)
         return;
-    }
 
-    std::optional<ScrollingNodeID> enclosingScrollingNodeID;
-    IntRect enclosingScrollingClipRect;
-    IntRect innerScrollingClipRect;
-    forEachEnclosingScroller(selection, [&](const std::optional<ScrollingNodeID>& scrollingNodeID, const IntRect& clipRectInRootView) {
-        if (scrollingNodeID) {
-            enclosingScrollingClipRect = clipRectInRootView;
-            enclosingScrollingNodeID = scrollingNodeID;
-            return IterationStatus::Done;
-        }
+    CheckedPtr enclosingLayer = startEnclosingLayer->commonAncestorWithLayer(*endEnclosingLayer);
+    if (!enclosingLayer)
+        return;
 
-        if (innerScrollingClipRect.isEmpty())
-            innerScrollingClipRect = clipRectInRootView;
+    for (; enclosingLayer; enclosingLayer = enclosingLayer->enclosingContainingBlockLayer(CrossFrameBoundaries::Yes)) {
```

Comment:
I think you can fold this together

```
for (CheckedPtr enclosingLayer = startEnclosingLayer->commonAncestorWithLayer(*endEnclosingLayer); enclosingLayer; enclosingLayer = enclosingLayer->enclosingContainingBlockLayer(CrossFrameBoundaries::Yes)) {
    ...
}
```

Reply:
(if you think this doesn't hurt readability)

Reply:
Done! (I'll line wrap the for loop increment logic)